### PR TITLE
Support custom layout call under torch dispatch mode

### PR DESCRIPTION
**Pull request by soulitzer**
Fixes #ISSUE_NUMBER

### CODE REVIEW

File: test/test_nestedtensor.py [hunk 1]
```diff
@@ -3876,6 +3876,14 @@ def test_is_contiguous(self, device):
         self.assertTrue(not nt_noncontiguous.is_contiguous(memory_format=torch.contiguous_format))
         self.assertTrue(nt_contiguous_narrow.is_contiguous(memory_format=torch.contiguous_format))
 
+    def test_layout_under_torch_dispatch_mode():
+        from torch.testing._internal.logging_tensor import capture_logs_with_logging_tensor_mode
+
+        nt = random_nt_from_dims([2, None, 3], torch.device('cpu'), torch.float32, layout=torch.jagged)
+
+        with capture_logs_with_logging_tensor_mode():
+            nt.layout
```
Comment (jbschlosser) [inline test/test_nestedtensor.py:3885]: nit: might as well check the value of layout as well :)

```python
self.assertEqual(nt.layout, torch.jagged)
```

File: test/test_nestedtensor.py [hunk 2]
```diff
@@ -3876,6 +3876,14 @@ def test_is_contiguous(self, device):
         self.assertTrue(not nt_noncontiguous.is_contiguous(memory_format=torch.contiguous_format))
         self.assertTrue(nt_contiguous_narrow.is_contiguous(memory_format=torch.contiguous_format))
 
+    def test_layout_under_torch_dispatch_mode():
```
Comment (jbschlosser) [inline test/test_nestedtensor.py:3879]:
```suggestion
    def test_layout_under_torch_dispatch_mode(self):
```

I think you need this for the test to run, no?

**Review by jbschlosser — COMMENTED** [review]
I'm still curious whether we can fix this in `prim.layout.default` or something, but the short-term fix seems okay

**Comment by soulitzer** [comment]
> I'm still curious whether we can fix this in prim.layout.default or something, but the short-term fix seems okay

Yup the root cause seems to be with `prim.layout.default` in that it goes through OpOverload which looks up and executes ops via the jit, and to the jit, Layout seems to be treated just like ints!

The issue is that when c10::Layout are converted IValue, it isn't actually tagged as a Layout and is simply converted to its underlying type. As a result, when the jit's wrapper later tries to convert the output stack back into PyObjects, it would not be able to determine whether a particular int is a IValue (unless we add custom logic to check the schema if any outputs are supposed to be Layout?). 

So, TLDR, the proper fix which is adding support to Layout in IValue seems doable, but will take some to iterate on ivalue.h (editing it will trigger many files to recompile). Wdyt about landing this short term fix, adding a comment, and leaving the long term fix for a follow up?

**Review by jbschlosser — APPROVED** [review]
> So, TLDR, the proper fix which is adding support to Layout in IValue seems doable, but will take some to iterate on ivalue.h (editing it will trigger many files to recompile). Wdyt about landing this short term fix, adding a comment, and leaving the long term fix for a follow up?

Proper fix seems like too much work; I'm good with this :) Thanks for looking into it though
